lib/dockerdaemon/cli.go

```diff
@@ -0,0 +1,182 @@
+func NewDockerClient(host, scheme, version string) (DockerClient, error) {
+	protocol, addr, basePath, err := parseHost(host)
+	if err != nil {
+		return nil, err
+	}
+
+	transport := new(http.Transport)
+	configureTransport(transport, protocol, addr)
+	client := &http.Client{
+		Transport: transport,
+	}
+
+	return &dockerClient{
+		scheme:   scheme,
+		host:     host,
+		version:  version,
+		protocol: protocol,
+		addr:     addr,
+		basePath: basePath,
+		client:   client,
+	}, nil
+}
+
+func parseHost(host string) (string, string, string, error) {
+	strs := strings.SplitN(host, "://", 2)
+	if len(strs) == 1 {
+		return "", "", "", fmt.Errorf("unable to parse docker host `%s`", host)
+	}
+
+	var basePath string
+	protocol, addr := strs[0], strs[1]
+	if protocol == "tcp" {
+		parsed, err := url.Parse("tcp://" + addr)
+		if err != nil {
+			return "", "", "", err
+		}
+		addr = parsed.Host
+		basePath = parsed.Path
+	}
+	return protocol, addr, basePath, nil
+}
+
+func configureTransport(tr *http.Transport, protocol, addr string) error {
+	switch protocol {
+	case "unix":
+		if len(addr) > maxUnixSocketPathSize {
+			return fmt.Errorf("Unix socket path %q is too long", addr)
+		}
+
+		tr.DisableCompression = true
+		tr.Dial = func(_, _ string) (net.Conn, error) {
+			return net.DialTimeout(protocol, addr, defaultTimeout)
+		}
+		return nil
+	}
+
+	return fmt.Errorf("Protocol %s not supported", protocol)
+}
+
+func (cli *dockerClient) post(
+	ctx context.Context, url string, query url.Values, body io.Reader,
+	header http.Header, streamRespBody bool) error {
+
+	if body == nil {
+		body = bytes.NewReader([]byte{})
+	}
+	resp, err := cli.doRequest(ctx, "POST", cli.getAPIPath(url, query), body, header)
+	if err != nil {
+		return fmt.Errorf("post request: %s", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		errMsg, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("read error resp: %s", err)
+		}
+		return fmt.Errorf("Error posting to %s: code %d, err: %s", url, resp.StatusCode, errMsg)
+	}
+
+	// Docker daemon returns 200 before complete push
+	// it closes resp.Body after it finishes
+	if streamRespBody {
+		log.Debugf("Streaming resp body for %s", url)
+		progress, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("read resp body: %s", err)
+		}
+		log.Debugf("%s", progress)
+	}
+
+	return nil
+}
+
+func (cli *dockerClient) getAPIPath(p string, query url.Values) string {
+	var apiPath string
+	if cli.version != "" {
+		v := strings.TrimPrefix(cli.version, "v")
+		apiPath = fmt.Sprintf("%s/v%s%s", cli.basePath, v, p)
+	} else {
+		apiPath = fmt.Sprintf("%s%s", cli.basePath, p)
+	}
+
+	u := &url.URL{
+		Path: apiPath,
+	}
+	if len(query) > 0 {
+		u.RawQuery = query.Encode()
+	}
+	return u.String()
+}
+
+func (cli *dockerClient) doRequest(
+	ctx context.Context,
+	method string,
+	url string,
+	body io.Reader,
+	header http.Header) (*http.Response, error) {
```

Comment:
The number of arguments for `post` and `doRequest` is a bit out of control, and as a result this file is a bit difficult to read. Can you think of a way to refactor this better? It seems like way too many knobs just to implement a single image pull http request.